c2corg_api/views/document_revert.py

```diff
@@ -0,0 +1,236 @@
+        raise HTTPBadRequest('Unknown version {}/{}/{}'.format(
+            document_id, lang, version_id))
+
+    # check the version to revert to is not the latest one
+    last_version_id, = DBSession.query(DocumentVersion.id). \
+        filter(and_(
+            DocumentVersion.document_id == document_id,
+            DocumentVersion.lang == lang)). \
+        order_by(DocumentVersion.id.desc()).first()
+    if version_id == last_version_id:
+        raise HTTPBadRequest(
+            'Version {}/{}/{} is already the latest one'.format(
+                document_id, lang, version_id))
+
+
+@resource(path='/documents/revert', cors_policy=cors_policy)
+class DocumentRevertRest(object):
+
+    def __init__(self, request):
+        self.request = request
+
+    @restricted_json_view(
+        permission='moderator',
+        schema=RevertSchema(),
+        validators=[colander_body_validator, validate_version])
+    def post(self):
+        """ Create a new version of the document based upon an old one.
+
+        Request:
+            `POST` `/documents/revert`
+
+        Request body:
+            {
+                'document_id': @document_id@,
+                'lang': @lang@,
+                'version_id': @version_id@
+            }
+
+        """
+        document_id = self.request.validated['document_id']
+        lang = self.request.validated['lang']
+        version_id = self.request.validated['version_id']
+
+        clazz, locale_clazz, archive_clazz, \
+            archive_locale_clazz, document_type = self._get_models(document_id)
+
+        document = self._get_current_document(
+            document_id, lang, clazz, locale_clazz)
+        document_in = self._get_archive_document(
+            document_id, lang, version_id, archive_clazz, archive_locale_clazz)
+
+        old_versions = document.get_versions()
+        user_id = self.request.authenticated_userid
+
+        # update the document with the input document
+        document.update(document_in)
+
+        # now set back the versions to the versions of the current document
+        # because we basing our changes on the lastest version
+        document.version = old_versions['document']
+        document.locales[0].version = old_versions['locales'][lang]
+        if document.geometry:
+            document.geometry.version = old_versions['geometry']
+
+        # TODO? support 'before_update' callbacks as in _put()
+        # See 'after_update' handling below
+
+        try:
+            DBSession.flush()
+        except StaleDataError:
+            raise HTTPConflict('concurrent modification')
+
+        # when flushing the session, SQLAlchemy automatically updates the
+        # version numbers in case attributes have changed. by comparing with
+        # the old version numbers, we can check if only figures or only locales
+        # have changed.
+        (update_types, changed_langs) = document.get_update_type(old_versions)
+
+        if update_types:
```

Comment:
This part looks very similar to the code in `DocumentRest._put`. Would it be possible to refactor a bit?